### ISSUE TITLE
Dotnetcore System.UInt32 conversion issues

### ISSUE THREAD

**Issue by jmlidbetter** [pythonnet/pythonnet#950]
### Environment

-   Pythonnet version: 2.4.0
-   Python version: 3.6.5
-   Operating System: Centos 7 64bit

### Details

-   Creating UInt32 in python with a value greater than 0x7FFFFFFF in dotnetcore will cause an overflow error. This is due (seemingly) to the fact that .NET core clr will cast the `uint` to a signed long (64 bits in size on my system) before calling the PyLong_FromUsignedLong function. Using GDB and breaking before the errornous overflow we can examine the created PyLongObject:

The `PyLongObject` given to C# by the python interpretter that can be converted to a `uint` using `PyLong_AsUnsignedLong`

![image](https://user-images.githubusercontent.com/53430310/64010679-2d37eb00-cb12-11e9-8323-3eb72be82f25.png)

Using `PyLong_FromUnsignedLong` on the resulting `uint` gives this:

![image](https://user-images.githubusercontent.com/53430310/64010712-3cb73400-cb12-11e9-96ec-3c7ae962433f.png)

I have raised an issue on the coreclr project: [Issue 26444](https://github.com/dotnet/coreclr/issues/26444)

FYI changing the function definition from
```C#
internal static extern IntPtr PyLong_FromUnsignedLong(uint value);
```
to
```C#
internal static extern IntPtr PyLong_FromUnsignedLong(ulong value);
```
fixes the issue.

**Comment by filmor** [comment]
Well, this is problematic. We want to be platform-independent, but C's `unsigned long` that is used in Python is 64bit on most 64bit Unices and 32bit on Windows, so the `uint` parameter is indeed wrong.

**Comment by Cronan** [comment]
I was wondering if there’s a Python C API call that might give us the unsigned king size at runtime, so we could decide which PInvoke call to make?

**Comment by filmor** [comment]
As far as I can tell, Python is completely oblivious to this difference as it "lives" on the C platform, so all compiled modules will have the same idea of what `long` means. We can pretty reliably decide this by looking at the platform we are running on in .NET, though. We are already providing internally booleans for `IsWindows` and `Is32Bit` here https://github.com/pythonnet/pythonnet/blob/master/src/runtime/runtime.cs#L106-L109.

**Comment by jmlidbetter** [comment]
> 
> 
> As far as I can tell, Python is completely oblivious to this difference as it "lives" on the C platform, so all compiled modules will have the same idea of what `long` means. We can pretty reliably decide this by looking at the platform we are running on in .NET, though. We are already providing internally booleans for `IsWindows` and `Is32Bit` here https://github.com/pythonnet/pythonnet/blob/master/src/runtime/runtime.cs#L106-L109.

Would this method get caught out if, for whatever reason, someone is using CPython compiled with 64bit longs on Win64 or CPython compiled with 32bit longs on 64bit Linux?

**Comment by jmlidbetter** [comment]
What we really need is some hook or proxy into `sizeof(long)`.

**Comment by Cronan** [comment]
As per a discussion with Joe, It seems the correct way to do this would be with a call to `struct.calcsize('P')` which can be done in `runtime.cs`, in `InitializePlatformData()`. This gives us the bitness of the interpreter architecture, rather than the platform architecture.

**Comment by filmor** [comment]
> > As far as I can tell, Python is completely oblivious to this difference as it "lives" on the C platform, so all compiled modules will have the same idea of what `long` means. We can pretty reliably decide this by looking at the platform we are running on in .NET, though. We are already providing internally booleans for `IsWindows` and `Is32Bit` here https://github.com/pythonnet/pythonnet/blob/master/src/runtime/runtime.cs#L106-L109.
> 
> Would this method get caught out if, for whatever reason, someone is using CPython compiled with 64bit longs on Win64 or CPython compiled with 32bit longs on 64bit Linux?

If you managed to compile a program that messed up, it will not be able to reliably interact with anything else on the same system. This is an ABI thing, it's not something a user can just decide.

**Comment by Cronan** [comment]
This is not about a user decision, it’s about calling the correct PInvoke signature at runtime, based on the bitness of the interpreter being run.

**Comment by filmor** [comment]
Yes, and the correct P/Invoke signature can be decided by looking at the `IsWindows` and `Is32Bit` values. You wouldn't be able to boot up a .NET VM (or load a `libpython` in case you're embedding) with non-matching bitness, so drawing those values from .NET will lead to the same result with much less work.

**Comment by jmlidbetter** [comment]
Fixed in PR #961
